### PR TITLE
EUSS Demand Flexibility Measures new lighting control measure and thermostat control updates

### PR DESCRIPTION
**Pull request by JieXiong9119**
Pull request overview
---------------------

Code review for new and updated demand flexibility measures and tests
 - lighting control load shed (new)
 - thermostat control load shed (updated)
 - thermostat control load shift (updated)
 - dispatch schedule generation (updated)

### Pull Request Author

This pull request makes changes to (select all the apply):
 - [ ] Documentation
 - [ ] Infrastructure (includes apptainer image, buildstock batch, dependencies, continuous integration tests)
 - [ ] Sampling
 - [ ] Workflow Measures
 - [x] Upgrade Measures
 - [ ] Reporting Measures
 - [ ] Postprocessing

Author pull request checklist:

 - [x] Tagged the pull request with the appropriate label (documentation, infrastructure, sampling, workflow measure, upgrade measure, reporting measure, postprocessing) to help categorize changes in the release notes.
 - [x] Added tests for new measures
 - [x] Updated measure .xml(s)
 - [ ] Register values added to `comstock_column_definitions.csv`
 - [x] Both `options_lookup.tsv` files updated
 - [x] 10k+ test run
 - [ ] Change documentation written
 - [x] Measure documentation written
 - [x] ComStock documentation updated
 - [x] Changes reflected in example `.yml` files
 - [ ] Changes reflected in `README.md` files
 - [x] Added 'See ComStock License' language to first two lines of each code file
 - [ ] Implements corresponding measure tests and indexing path in `test/measure_tests.txt` or/and `test/resource_measure_tests.txt`
 - [ ] All new and existing tests pass the CI

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a code review on GitHub
 - [ ] All related changes have been implemented: data and method additions, changes, tests
 - [ ] If fixing a defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] Reviewed change documentation
 - [ ] Ensured code files contain License reference
 - [ ] Results differences are reasonable
 - [ ] Make sure the newly added measures has been added with tests and indexed properly
 - [ ] CI status: all tests pass

#### ComStock Licensing Language - Add to Beginning of Each Code File
```
# ComStock™, Copyright (c) 2023 Alliance for Sustainable Energy, LLC. All rights reserved.
# See top level LICENSE.txt file for license terms.
```

### CODE REVIEW

File: resources/measures/upgrade_df_lighting_control/tests/df_lighting_control_test.rb [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    paths = paths.map { |path| File.expand_path(path) }
+    return paths
+  end
+
+  # create an array of hashes with model name, weather, and expected result
+  def models_to_test
+    test_sets = []
+
+    # test: applicable building type
+    test_sets << {
```
Comment (ChristopherCaradonna) [inline resources/measures/upgrade_df_lighting_control/tests/df_lighting_control_test.rb]: you should replace the example models with the most recent 3.7 versions.

File: resources/measures/upgrade_df_thermostat_control_load_shed/measure.rb [hunk 2]
```diff
@@ -84,12 +84,23 @@ def arguments(model)
     num_timesteps_in_hr.setDefaultValue(4)
     args << num_timesteps_in_hr
 
-    choices = ['full baseline', 'bin sample', 'part year bin sample']
+    choices = ['full baseline', 'bin sample', 'part year bin sample', 'fix', 'oat']
```
Comment (ChristopherCaradonna) [inline resources/measures/upgrade_df_thermostat_control_load_shed/measure.rb]: Do you have the measure options described in the measure or parameter descriptions?

File: resources/measures/upgrade_df_thermostat_control_load_shift/measure.rb [hunk 3]
```diff
@@ -461,12 +482,15 @@ def applicable_thermostats(model)
...
-      # "MediumOffice",
+      "SmallOffice",
+      "MediumOffice",
       "LargeOffice",
       "OfL",
       "OfS",
-      "Office"
+      "Office",
+      "Warehouse",
+      "PrimarySchool",
```
Comment (ChristopherCaradonna) [inline resources/measures/upgrade_df_thermostat_control_load_shift/measure.rb]: are there space types in schools that you are excluding? Is it reasonable to apply these to gyms, cafeteria, etc.?

File: resources/tests/models/3340.osm [hunk 4]
```diff
@@ -0,0 +1,12463 @@
+
```
Comment (ChristopherCaradonna) [inline resources/tests/models/3340.osm]: Can you save these test models with a more descriptive name, including the OS version?

**Review by ChristopherCaradonna — COMMENTED** [review]
Some minor comments within code. Do the measure tests check for more than just success? If not, it would be good to add a test that checks for proper application. Additionally, add a link to the single model example when it's ready.

File: resources/measures/upgrade_df_lighting_control/measure.rb [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    peak_lag = runner.getIntegerArgumentValue("peak_lag",user_arguments)
+    peak_window_strategy = runner.getStringArgumentValue("peak_window_strategy",user_arguments)
+    # apply_measure = runner.getBoolArgumentValue('apply_measure', user_arguments)
+
+    def light_adj_based_on_sch(peak_sch, light_adjustment)
+      light_adj_values = peak_sch.map{|a| 1-light_adjustment/100.0*a}
+      return light_adj_values
+    end
+    
+    def get_hourly_schedule_from_schedule_ruleset(model, schedule_ruleset)
```
Comment (eringold) [inline resources/measures/upgrade_df_lighting_control/measure.rb]: Look at using this OpenstudioStandards::Schedule method: https://github.com/NREL/openstudio-standards/blob/8f948207d6af73165a2a8232559804b93d8e50c2/lib/openstudio-standards/schedules/information.rb#L106

File: resources/measures/upgrade_df_lighting_control/measure.rb [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    ############################################
+    # Generate peak schedule
+    ############################################
+    puts("### ============================================================")
+    puts("### Creating peak schedule...")
+    if load_prediction_method == 'fix'
+      puts("### Fixed schedule...")
+      template = 'ComStock 90.1-2019'
+      std = Standard.build(template)
+      climate_zone = std.model_standards_climate_zone(model)
```
Comment (eringold) [inline resources/measures/upgrade_df_lighting_control/measure.rb]: This is now `OpenstudioStandards::Weather.model_get_climate_zone`, with no need to build a Standard class: https://github.com/NREL/openstudio-standards/blob/8f948207d6af73165a2a8232559804b93d8e50c2/lib/openstudio-standards/weather/information.rb#L92

**Review by eringold — COMMENTED** [review]
FYI

File: resources/measures/upgrade_df_lighting_control/measure.rb (hunk 5 above)
Comment (JieXiong9119) [reply resources/measures/upgrade_df_lighting_control/measure.rb]: Is there a way to choose sub-hourly interval for this method? I name the function "hourly" but it actually outputs hourly or 15min values.

Comment (eringold) [reply resources/measures/upgrade_df_lighting_control/measure.rb]: That method will only output hourly (8760) values. To get timestep values, you could replace almost everything in the `day_sched_array.each` block with this new (in 3.8.0) method [`ScheduleDay::timeSeries`](https://s3.amazonaws.com/openstudio-sdk-documentation/cpp/OpenStudio-3.8.0-doc/model/html/classopenstudio_1_1model_1_1_schedule_day.html#a063fc4b5d4b6cc6979ce732e9bc314c7).

File: resources/measures/upgrade_df_thermostat_control_load_shift/measure.rb (hunk 3 above)
Comment (JieXiong9119) [reply resources/measures/upgrade_df_thermostat_control_load_shift/measure.rb]: We assume a uniform applicability to all space type to simplify the measure (for now) with the following considerations:
1. The control is applicable to the majority of spaces in the building;
2. The overestimated potential of applying the measure would compensate the underestimated potentia - for example, the lights in corridor area could be dimmed more while the lights in gym should be dimmed less (or skipped) - and such trade-offs are assumed to have negligible impact on the overall performance (especially in stock level).

These are explained in the documentation as well (in the applicability section). Let me know your opinion on this.

**Review by eringold — APPROVED** [review]

**Pull request merged**